### PR TITLE
Document remaining message types in SIM-PROTOCOL

### PR DESCRIPTION
**Pull request by scheibo**
Oh yeah, a sexy and glamorous documentation PR!

Think this is now exhaustive from grepping through the client/server source (well, maybe `-anim`, but that seems OM only?). The client seems to generate its own artificial events (`-block`, etc) which are obviously not covered here. Had trouble coming up with definitions for some of these, so this PR is WIP until I figure them out or someone chimes in with them. :)

- [x] not fully clear on why `upkeep` is required (does `turn` not suffice?)
- [x]  `-start` and `-end` seems a bit complex and I wanted some guidance on what to write?
- [x]  still need to document `-singlemove` and `singleturn`, maybe could use some wording help here

### CODE REVIEW

**Comment by scheibo** [comment]
Hmm, `-singleturn` seems to use both move *names* and moves in the *effect* format:

```
...
data/moves.js:                            this.add('-singleturn', source, 'Quick Guard');
data/moves.js:                            this.add('-singleturn', pokemon, 'move: Rage Powder');
data/moves.js:                            this.add('-singleturn', target, 'move: Roost');
data/moves.js:                            this.add('-singleturn', pokemon, 'move: Shell Trap');
data/moves.js:                            this.add('-singleturn', pokemon, 'Snatch');
...
```

is this deliberate?

**Comment by scheibo** [comment]
`upkeep` is still kind of question, but `-start` and `-end` just seem to be volatile statuses (doh). I think this is ready for review.

**Comment by Zarel** [comment]
`-block` is the only artificial message type, which is planned but not yet implemented in the server.

File: sim/SIM-PROTOCOL.md [hunk 1]
```diff
@@ -443,10 +523,41 @@ stat boosts are minor actions.
...
+> The `ATTACKER` Pokémon is preparing to use a charge `MOVE` on the `DEFENDER`
+> (For example: Dig, Fly).
+
+`|-mustrecharge|POKEMON`
+
+> The Pokémon `POKEMON` must spend the turn recharging from a previous move.
+
+`|-nothing`
+
+> A move did absolutely nothing. (For example: Splash).
```
Comment (Zarel) [inline sim/SIM-PROTOCOL.md]: This is deprecated; it's now `|-activate|...|move: Splash`.

File: sim/SIM-PROTOCOL.md [hunk 2]
```diff
@@ -443,10 +523,41 @@ stat boosts are minor actions.
...
-the server on a regular connection, or look at the client source code
-for a full list of message types.
+`|-combine`
+
+> A move has been combined with another (For example: Fire Pledge).
+
+`|-waiting|SOURCE|TARGET`
+
+> The `SOURCE` Pokémon has used a move and is waiting for the `TARGET` Pokémon
+> (For example: Fire Pledge).
```
Comment (Zarel) [inline sim/SIM-PROTOCOL.md]: I think this is also deprecated for `|-activate|` of the relevant Pledge move.

File: sim/SIM-PROTOCOL.md (hunk 1 above)
Comment (scheibo) [reply sim/SIM-PROTOCOL.md]: Yeah, I noticed that on the client the `upgradeArgs`, but it still seems to be generated by the server? e.g. https://github.com/Zarel/Pokemon-Showdown/blob/master/data/mods/gen2/moves.js#L736

Should I remove this entitely, or just add a note in the documentation that it's deprecated?

File: sim/SIM-PROTOCOL.md (hunk 2 above)
Comment (scheibo) [reply sim/SIM-PROTOCOL.md]: https://github.com/Zarel/Pokemon-Showdown/blame/master/data/moves.js#5137 (and elsewhere?)

File: sim/SIM-PROTOCOL.md (hunk 1 above)
Comment (Zarel) [reply sim/SIM-PROTOCOL.md]: Either is up to you, but `upgradeArgs` is explicitly supposed to upgrade a deprecated message type into the modern equivalent. Anything sent by the server that's handled by `upgradeArgs` is supposed to eventually become the new standard sent by the server.

Comment (Zarel) [reply sim/SIM-PROTOCOL.md]: Although with some extreme caveats that `upgradeArgs` represents "proposed" changes in the sense that I don't want to go through and fix server output _yet_.

Comment (Zarel) [reply sim/SIM-PROTOCOL.md]: So marking it as "deprecated" here may be the best solution.

File: sim/SIM-PROTOCOL.md (hunk 2 above)
Comment (Zarel) [reply sim/SIM-PROTOCOL.md]: I guess the Pledge move messages are currently a huge hack that I need to go through and upgrade at some point.

**Comment by Zarel** [comment]
`-upkeep` is used by `updatePseudoWeatherLeft`, to update the number of turns of field conditions left (weather, terrain, etc). I think the idea is to count turns correctly during a death-switch.

**Pull request merged**